google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/RateLimitingServerStreamingCallable.java

```diff
@@ -0,0 +1,149 @@
+/*
+ * Copyright 2023 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.cloud.bigtable.data.v2.stub;
+
+import com.google.api.gax.rpc.ApiCallContext;
+import com.google.api.gax.rpc.DeadlineExceededException;
+import com.google.api.gax.rpc.ResourceExhaustedException;
+import com.google.api.gax.rpc.ResponseObserver;
+import com.google.api.gax.rpc.ServerStreamingCallable;
+import com.google.api.gax.rpc.StreamController;
+import com.google.api.gax.rpc.UnavailableException;
+import com.google.bigtable.v2.MutateRowsRequest;
+import com.google.bigtable.v2.MutateRowsResponse;
+import com.google.bigtable.v2.RateLimitInfo;
+import com.google.cloud.bigtable.data.v2.stub.metrics.BigtableTracer;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+import com.google.common.base.Stopwatch;
+import com.google.common.util.concurrent.RateLimiter;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.logging.Logger;
+import javax.annotation.Nonnull;
+import org.threeten.bp.Duration;
+
+class RateLimitingServerStreamingCallable
+    extends ServerStreamingCallable<MutateRowsRequest, MutateRowsResponse> {
+  private static final Logger logger =
+      Logger.getLogger(RateLimitingServerStreamingCallable.class.getName());
+  private static final long DEFAULT_QPS = 10;
+  // Default interval before changing the QPS on error response
+  private static Duration DEFAULT_PERIOD = Duration.ofSeconds(10);
+  private final RateLimiter limiter;
+  private final AtomicLong lastQpsChangeTime = new AtomicLong(System.currentTimeMillis());
+  private final ServerStreamingCallable<MutateRowsRequest, MutateRowsResponse> innerCallable;
+  private static final double MIN_QPS = 0.1;
+  private static final double MAX_QPS = 100_000;
+  @VisibleForTesting static final double MIN_FACTOR = 0.7;
+  private static final double MAX_FACTOR = 1.3;
```

Comment:
Also please comment what they are used for. ie MIN_FACTOR is used as the assumed QPS scaling factor when an RPC ends in an error which doesn't have the trailer